web/src/queries/zfcp.ts

```diff
@@ -0,0 +1,179 @@
+ *
+ * All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as published
+ * by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+import { useMutation, useQuery, useQueryClient, useSuspenseQuery } from "@tanstack/react-query";
+import { _ } from "~/i18n";
+import { fetchZFCPConfig, fetchZFCPControllers, fetchZFCPDisks } from "~/api/zfcp";
+import { useInstallerClient } from "~/context/installer";
+import React from "react";
+import { ZFCPConfig, ZFCPController, ZFCPDisk } from "~/types/zfcp";
+
+/**
+ * Returns a query for retrieving the zFCP controllers
+ */
+const ZFCPControllersQuery = () => ({
+  queryKey: ["zfcp", "controllers"],
+  queryFn: fetchZFCPControllers,
+});
+
+/**
+ * Returns a query for retrieving the zFCP disks
+ */
+const ZFCPDisksQuery = () => ({
+  queryKey: ["zfcp", "disks"],
+  queryFn: fetchZFCPDisks,
+});
+
+/**
+ * Returns a query for retrieving the zFCP config
+ */
+const ZFCPConfigQuery = () => ({
+  queryKey: ["zfcp", "config"],
+  queryFn: fetchZFCPConfig,
+});
+
+/**
+ * Hook that returns zFCP controllers.
+ */
+const useZFCPControllers = (): ZFCPController[] => {
+  const { data: controllers } = useSuspenseQuery(ZFCPControllersQuery());
+  return controllers;
+};
+
+/**
+ * Hook that returns zFCP disks.
+ */
+const useZFCPDisks = (): ZFCPDisk[] => {
+  const { data: devices } = useSuspenseQuery(ZFCPDisksQuery());
+  return devices;
+};
+
+/**
+ * Hook that returns zFCP config.
+ */
+const useZFCPConfig = (): ZFCPConfig => {
+  const { data: config } = useSuspenseQuery(ZFCPConfigQuery());
+  return config;
+};
+
+/**
+ * Listens for zFCP Controller changes.
+ */
+const useZFCPControllersChanges = () => {
+  const client = useInstallerClient();
+  const queryClient = useQueryClient();
+
+  React.useEffect(() => {
+    if (!client) return;
+
+    return client.ws().onEvent((event) => {
+      switch (event.type) {
+        case "ZFCPControllerAdded": {
+          const device: ZFCPController = event.device;
+          queryClient.setQueryData(["zfcp", "controllers"], (prev: ZFCPController[]) => {
+            return [...prev, device];
+          });
+          break;
+        }
+        case "ZFCPControllerRemoved": {
+          const device: ZFCPController = event.device;
+          const { id } = device;
+          queryClient.setQueryData(["zfcp", "controllers"], (prev: ZFCPController[]) => {
```

Comment:
If you write the code in a slightly different way, you can make it simpler. Given that `setQueryData` receives a function, why not writing the code the other way around?

```javascript
queryClient.setQueryData(ZFCPConfigQuery.queryKey, (prev: ZFCPController[] | undefined) => {
  if (prev === undefined) return;

  // Put the switch here
})
```

Please be aware that `prev` might be `undefined`. Returning `undefined` from the updater function would make the query to not be updated. Similar to iSCSI [useNodesChanges](https://github.com/openSUSE/agama/pull/1598/files#diff-238cd70d57b6ea7dc5fa5411500ee14efca3f66dd775be6693f64e5dfb735126R96).

Reply:
agreed, this is mainly my attempt to have it idempotent and your code do it in easier way.